### ISSUE TITLE
Preserve branch position when going up and down

### ISSUE THREAD

**Issue by countvajhula** [countvajhula/symex.el#17]
When going up and down the symex, it would be nice to preserve the branch position so that up/down become inverse operations.

E.g. If we're on symex index 3 and go down, then coming back up should return to symex 3, since it is an arbitrary choice in any case, and is intuitive since it's the inverse of going down and returns us to our original position higher up in the tree.

### Options

**A**. It may make sense to wait for [the AST](https://github.com/countvajhula/symex.el/issues/3) on this one, since that would be the most reliable and simplest way to do this, but #3 is a non-trivial effort
**B**. A possible implementation in the meantime, which would add some maintenance overhead but could be simple to implement:
  - if `up` succeeds, increment height
  - if `down` succeeds, check if at root and set to `0` (as a failsafe), otherwise decrement height
  - before going `down`, store the current index in a hash, `{height: index}`
  - on going `up`, after incrementing height, check if a stored position exists for the new height in the hash. If it does, then `(symex-go-forward count)` that many times
  - on `left` and `right`, if at root, clear the height hash since it's only relevant for the original tree

Potential issues with **Option B** (would not be issues with AST, option A):

- on `swallow` operations, the heights are offset by one higher up from that point, but remain the same lower down
- `substitute`/ `change` / `delete` may also need special handling to note changed offsets
- `emit` and `capture` too, but they should do the expected thing on failure (e.g stop short if the original symex isn't available anymore)
- `raise`

Basically any transformation of the symex would need to be reviewed. We could also just blanket flush the hash on any transformation, which may be the simplest thing to do.

**Comment by countvajhula** [comment]
Added a simple implementation using a stack.
